lib/linked_list.rb

```diff
@@ -14,92 +14,193 @@ def initialize(value, next_node = nil)
+          pointer = pointer.next
+      end 
+
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), where n is the index given
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if @head == nil
+      return @head.data if index == 0
+
+      pointer = @head 
+      count = 0
+
+      until count == index
+          count += 1
+          pointer = pointer.next
+      end 
+
+      return pointer.data  
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), where n is the dependant on the length of the list
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      return nil if @head == nil
+      print @head.data if @head.next == nil
+
+      pointer = @head 
+
+      while pointer
+          print pointer.data
+          pointer = pointer.next
+      end 
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), where n is the dependant on the length of the list
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return if !@head
+      @head = @head.next if @head.data == value
+      pointer = @head
+      previous = nil
+
+      #go through each node
+      while pointer
+      #if the node equals the value the previous node.next should pointer to current node.next and the next node.previous needs to equal the previous node
+        if pointer.data == value
+          previous.next = pointer.next if previous
+          pointer = nil
+          return
+        else
+          previous = pointer 
+          pointer = pointer.next   
+        end
+      end
+      return
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), where n is the dependant on the length of the list
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      return nil if @head.nil?
+      return @head if @head.next.nil?
+      pointer = @head
+      previous = nil
+      #move each node.next to previous
+      while pointer
+        temp = previous
+        previous = pointer
+        pointer = pointer.next
+        previous.next = temp
+      end
+      # Reset the head to the new start
+      @head = previous
     end
 
+    ## ---
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def find_middle_value

```

Comment:
⚠️ This returns the index number of the middle and not the middle element of the list.